radixdlt-core/radixdlt/src/integration/java/com/radixdlt/integration/serialization/MessageCentralFuzzyTest.java

```diff
@@ -0,0 +1,127 @@
+import com.radixdlt.utils.Compress;
+
+import java.security.SecureRandom;
+import java.util.Comparator;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicLong;
+
+import io.reactivex.rxjava3.subjects.PublishSubject;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MessageCentralFuzzyTest {
+	private static final int MIN_MESSAGE_LEN = 1;
+	private static final int MAX_MESSAGE_LEN = 1024 * 1024;
+	private static final int NUM_TEST_MESSAGES = 1000;
+
+	private final Random random = new SecureRandom();
+	private final Serialization serialization = DefaultSerialization.getInstance();
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void fuzzy_messaged_are_not_accepted() throws Exception {
+		var inboundMessages = PublishSubject.<InboundMessage>create();
+		var config = mock(MessageCentralConfiguration.class);
+		var peerControl = mock(PeerControl.class);
+		var peerManager = mock(PeerManager.class);
+		var queueFactory = mock(EventQueueFactory.class);
+
+		when(config.messagingOutboundQueueMax(anyInt())).thenReturn(1);
+		when(config.messagingTimeToLive(anyLong())).thenReturn(30_000L);
+		when(peerManager.messages()).thenReturn(inboundMessages);
+
+		when(queueFactory.createEventQueue(anyInt(), any(Comparator.class)))
+			.thenReturn(new SimplePriorityBlockingQueue<>(1, OutboundMessageEvent.comparator()));
+
+		var messageCentral = new MessageCentralImpl(
+			config,
+			serialization,
+			peerManager,
+			Time::currentTimestamp,
+			queueFactory,
+			new SystemCountersImpl(),
+			() -> peerControl
+		);
+
+		var counter = new AtomicLong(0);
+
+		var disposable = messageCentral.messagesOf(Message.class)
+			.subscribe(nextItem -> counter.incrementAndGet(), error -> fail(error.getMessage()));
+
+		emitFuzzyMessages(inboundMessages);
+		disposable.dispose();
+
+		assertEquals(1L, counter.get());
+	}
+
+	private void emitFuzzyMessages(PublishSubject<InboundMessage> emitter) {
+		try {
+			//Insert single valid message to ensure whole pipeline is working properly.
+			var bytes = Compress.compress(serialization.toDson(new PeerPingMessage(), DsonOutput.Output.WIRE));
+			var valid = new InboundMessage(Time.currentTimestamp(), randomNodeId(), bytes);
+			emitter.onNext(valid);
+		} catch (Exception e) {
+			// Ignore
+		}
+
+		// Insert batch of randomly generated messages
+		for (int i = 0; i < NUM_TEST_MESSAGES; i++) {
+			emitter.onNext(generateFuzzyMessage());
+		}
+
+		emitter.onComplete();
+	}
+
+	private InboundMessage generateFuzzyMessage() {
+		while (true) {
+			try {
+				var compressedMessage = Compress.compress(generateRandomBytes());
+				return new InboundMessage(Time.currentTimestamp(), randomNodeId(), compressedMessage);
```

Comment:
I was imagining, as well as a few random-byte tests, we could employ some more structured fuzzing of the Java objects which get compressed -- something like: https://github.com/rohanpadhye/JQF - which is much more likely to catch bugs.

What do you think?

Reply:
Okay, so the intention of this test is to ensure the MessagePreprocessor doesn't crash on semantically invalid inputs? I'm glad it caught some issues 👍 

I still think there is a lot of benefit in automatically generating semantically valid messages in various forms and feeding them in, and checking for exceptions, as it means we're less likely to miss cases. But those may need to be fed through the whole app to be a sensible test I imagine?

Reply:
Yes, and to cover all parts which are involved in the message processing, such a test should be done at the system level.